### PR TITLE
maddy 1.1.2 recipe

### PR DESCRIPTION
**Pull request by sebastianbergt**
Specify library name and version:  **maddy/1.1.2**

I needed a c++ markdown parser for a project of mine and could not find one within CCI.
It is not a dependency for other libraries I want to use. I am not the author of the library.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/maddy/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import os
+from conans import ConanFile, tools
+
+
+class maddyConan(ConanFile):
```
Comment (madebr) [inline recipes/maddy/all/conanfile.py:5]:
```suggestion
from conans import ConanFile, tools
import os

required_conan_version = ">=1.33.0"


class MaddyConan(ConanFile):
```

File: recipes/maddy/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    no_copy_source = True
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename("maddy-{}".format(self.version), self._source_subfolder)
```
Comment (madebr) [inline recipes/maddy/all/conanfile.py:21]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version],
                  destination=self_source_subfolder, strip_root=True)
```

File: recipes/maddy/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.5)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(example example.cpp)
+target_link_libraries(example ${CONAN_LIBS})
+set_target_properties(example PROPERTIES CXX_STANDARD 14 CXX_STANDARD_REQUIRED
+                                                         ON)
```
Comment (madebr) [inline recipes/maddy/all/test_package/CMakeLists.txt:10]:
```suggestion
conan_basic_setup(TARGETS)

find_package(maddy REQUIRED CONFIG)

add_executable(${PROJECT_NAME} example.cpp)
target_link_libraries(${PROJECT_NAME} PRIVATE maddy::maddy)
set_target_properties(${PROJECT_NAME} PROPERTIES CXX_STANDARD 14 CXX_STANDARD_REQUIRED
                                                                 ON)
```

File: recipes/maddy/all/test_package/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class MaddyTestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
```
Comment (madebr) [inline recipes/maddy/all/test_package/conanfile.py:7]:
```suggestion
class TestPackageConan(ConanFile):
    settings = "os", "arch", "compiler", "build_type"
    generators = "cmake", "cmake_find_package_multi"
```

File: recipes/maddy/all/test_package/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self.settings):
```
Comment (madebr) [inline recipes/maddy/all/test_package/conanfile.py:15]:
```suggestion
        if not tools.cross_building(self):
```

File: recipes/maddy/all/test_package/example.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/*
+ *  This example was copied from https://github.com/progsource/maddy/blob/master/README.md
+ */
+
+#include <memory>
```
Comment (madebr) [inline recipes/maddy/all/test_package/example.cpp:5]: My suggestions to the test package are completely superfluous, but it's nice to see some output imho.
My changes are small.
```suggestion
#include <iostream>
#include <memory>
```

File: recipes/maddy/all/test_package/example.cpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+#include <string>
+#include <cstdlib>
+
+#include <maddy/parser.h>
+
+int main()
+{
+   std::stringstream markdownInput("");
```
Comment (madebr) [inline recipes/maddy/all/test_package/example.cpp:13]:
```suggestion
   std::stringstream markdownInput("#Hello world from maddy!\nVisit [conan-center-index](https://github.com/conan-io/conan-center-index)\n");
```

File: recipes/maddy/all/test_package/example.cpp [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+   // config is optional
+   std::shared_ptr<maddy::ParserConfig> config = std::make_shared<maddy::ParserConfig>();
+   config->isEmphasizedParserEnabled = true; // default
+   config->isHTMLWrappedInParagraph = true;  // default
+
+   std::shared_ptr<maddy::Parser> parser = std::make_shared<maddy::Parser>(config);
+   std::string htmlOutput = parser->Parse(markdownInput);
```
Comment (madebr) [inline recipes/maddy/all/test_package/example.cpp:21]:
```suggestion
   std::string htmlOutput = parser->Parse(markdownInput);

   std::cout << "html:\n" << htmlOutput;
```

File: recipes/maddy/all/test_package/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.5)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
```
Comment (madebr) [inline recipes/maddy/all/test_package/CMakeLists.txt:5]:
```suggestion
conan_basic_setup(TARGETS)
```

File: recipes/maddy/all/test_package/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+            bin_path = os.path.join("bin", "example")
```
Comment (madebr) [inline recipes/maddy/all/test_package/conanfile.py:16]:
```suggestion
            bin_path = os.path.join("bin", "test_package")
```

File: recipes/maddy/all/test_package/CMakeLists.txt (hunk 9 above)
Comment (sebastianbergt) [reply recipes/maddy/all/test_package/CMakeLists.txt]: done, thank you for your feedback :)

File: recipes/maddy/all/test_package/example.cpp [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include <maddy/parser.h>
+
+int main() {
+  std::stringstream markdownInput(
+      "#Hello world from maddy!\nVisit "
```
Comment (madebr) [inline recipes/maddy/all/test_package/example.cpp:15]: Looks like there must be a space after the hashtag.
```suggestion
      "# Hello world from maddy!\nVisit "
```

File: recipes/maddy/all/test_package/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
+
+find_package(maddy REQUIRED CONFIG)
+
+add_executable(${PROJECT_NAME} example.cpp)
+target_link_libraries(${PROJECT_NAME} PRIVATE maddy::maddy)
+set_target_properties(${PROJECT_NAME} PROPERTIES CXX_STANDARD 14 CXX_STANDARD_REQUIRED
```
Comment (prince-chrismc) [inline recipes/maddy/all/test_package/CMakeLists.txt:11]: Some validate logic is missing here :thinking:

**Review by prince-chrismc — COMMENTED** [review]
Since this project requires C++14 we need to make sure the compiler and settings are configured correctly.

conan-io/conan#8002 is my usual go to for copy pasting

You can also look at the SML recipe for an example, I just opened a pr to open it for a qucker link #8048

**Comment by sebastian-bergt** [comment]
Sorry for the delay. Upon closer inspection the code of the library does not satisfy my quality criteria. PR can be closed.

File: recipes/maddy/all/test_package/CMakeLists.txt (hunk 12 above)
Comment (SSE4) [reply recipes/maddy/all/test_package/CMakeLists.txt:11]: we don't have any validate logic in the CMakeLists.txt
we may add `validate` method to the recipe, if library requires C++14 to build.
but if test package requires C++14 standard, it's another story.

**Review by SSE4 — DISMISSED** [review]

File: recipes/maddy/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    )
+    topics = ("maddy", "markdown", "header-only")
+    license = "MIT"
+
+    no_copy_source = True
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
```
Comment (SpaceIm) [inline recipes/maddy/all/conanfile.py:22]:
```suggestion

    def validate(self):
        if self.settings.compiler.get_safe("cppstd"):
            tools.check_min_cppstd(self, 14)

```

It's a C++14 library.

File: recipes/maddy/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+class MaddyConan(ConanFile):
+    name = "maddy"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/progsource/maddy"
+    description = (
+        "open-source, maddy is a C++ Markdown to HTML header-only parser library."
+    )
+    topics = ("maddy", "markdown", "header-only")
+    license = "MIT"
+
```
Comment (SpaceIm) [inline recipes/maddy/all/conanfile.py:16]:
```suggestion
    settings = "os", "arch", "compiler", "build_type"
```

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**